rfcs/45-stageless.md

```diff
@@ -0,0 +1,727 @@
+   7. Use [system builder syntax](https://github.com/bevyengine/rfcs/pull/31), rather than adding more complex methods to `App`
+5. Re-add strict ordering constraints
+6. Add run criteria
+   1. Create the machinery to generate run criteria from functions
+   2. Cache the accesses of run criteria on `ConfiguredSystem` in the schedule
+   3. Check the values of the run-criteria of systems before deciding whether it can be run
+7. Add states
+   1. Create `State` trait
+   2. Implement while-active states as sugar for simple run criteria
+   3. Create on-enter and on-exit schedules
+   4. Create sugar for adding systems to these schedules
+8. Rename "system chaining" to "system piping"
+   1. Usage was very confusing for new users
+   2. New concept of "systems with a linear graph of ordering constraints between them" is naturally described as a chain
+9. Add new examples
+    1. Complex control flow with supplementary schedules
+    2. Fixed time-step pattern
+10. Port the entire engine to the new paradigm
+    1. We almost certainly need to port the improved ambiguity checker over to make this new paradigm usable
+
+Given the massive scope, that sounds relatively straightforward!
+However, doing so will break approximately the entire engine, and tests will not pass again until step 10.
+
+Lets explore some of the more critical and challenging details.
+
+### Storing and configuring systems
+
+Schedules store systems in a configured, initialized form.
+
+```rust
+struct Schedule{
+   // Each schedule is associated with exactly one `World`, as systems must be initiliazed on it.
+
+   world_id: WorldId
+   // We need to be able to quickly look up specific systems and iterate over them.
+   // `SystemId` should be a unique opaque identifier, generated on system insertion.
+   // We cannot use a Vec + a `usize` index as the identifier,
+   // as it would not be stable when systems were removed, risking serious bugs.
+   systems: HashMap<SystemId, ConfiguredSystem>,
+   // Stores the fully initialized, efficient graph of ordering constraints
+   // for all systems in the schedule.
+   // The properties of each set are passed down to specific systems to cache computation results.
+   // `Graph` is a bikeshed-avoidance graph storage structure,
+   // the exact strategy should be benchmarked.
+   ordering_constraints: Graph<SystemId>,
+   // Set configuration must be stored at the schedule level,
+   // rather than attached to specific set instances.
+   sets: HashMap<Box<dyn SystemLabel>, SystemSetConfig>,
+   // Used to check if mutation is allowed.
+   currently_running: bool,
+}
+```
+
+Configured systems are composed of a raw system and its configuration.
+
+```rust
+struct ConfiguredSystem{
+   raw_system: System,
+   config: SystemConfig,
+}
+```
+
+Raw systems store metadata, cached state, `Local` system parameters and the function pointer to the actual function to be executed:
+
+```rust
+struct System {

```

Comment:
Can you add full types here and propagate them up where relevant? As written this won't work because System is generic on F and ConfiguredSystem cannot be generic (and will require boxing the system). It also seems like this is advocating for removing the system trait in favor of a struct? Imo removing the System trait is _very_ opinionated and deserves its own RFC (or at the very least a section in this one).